### PR TITLE
Add optional `step_hook` to SFT trainer to enable extra training-time logic

### PR DESCRIPTION
**Pull request by karinazad**
There was a recent [reverted PR](https://github.com/PrimeIntellect-ai/prime-rl/pull/1709) for adding online validation support in SFT training. In trying to add similar validation behavior in our integration, we ended up maintaining a full copied version of train.py just to inject validation loops. That approach works short-term, but it is difficult to keep aligned with upstream trainer changes over time.
This PR proposes a more general extension point so validation or other training-time logic can be implemented externally without changing or forking the core loop each time.

Changes:
- Adds an optional `step_hook` argument to `prime_rl.trainer.sft.train.train(...)`
- Adds `SFTStepHookContext` with core step state (config, model, monitor, progress, loss object, CP context)
- Calls the hook once per training step when provided
- Leaves behavior unchanged when no hook is passed

Simplified example of usage:

```py
from prime_rl.trainer.sft.train import SFTStepHookContext, train

def custom_step_hook(ctx: SFTStepHookContext) -> None:
    step = ctx.progress.step

    # Put any custom per-step logic here.
    # Possible use cases:
    # - periodic validation / perplexity logging
    # - custom experiment metrics
    # - adaptive curriculum or data-mixture updates
    # - early-stop / plateau signals
    # - lightweight debugging or profiling checks
    custom_metrics = run_custom_logic(
        model=ctx.model,
        ce_loss=ctx.ce_loss,
        cp_enabled=ctx.cp_enabled,
        cp_rank=ctx.cp_rank,
        cp_size=ctx.cp_size,
        cp_group=ctx.cp_group,
        step=step,
    )

    if custom_metrics:
        custom_metrics["step"] = step
        ctx.monitor.log(custom_metrics, step=step)


train(config, step_hook=custom_step_hook)
```

---

> [!NOTE]
> **Medium Risk**
> Adds a new extension point that runs user-provided code every training step and changes the `train(...)` function signature, which can impact performance or introduce side effects when used. Core training behavior should be unchanged when `step_hook` is not provided.
> 
> **Overview**
> **Adds an extension point to SFT training.** `train(...)` now accepts an optional `step_hook` callback.
> 
> Introduces `SFTStepHookContext` (config/model/monitor/progress/loss + CP info) and invokes the hook once per step after metrics logging; behavior remains the same when no hook is passed.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit e2d17291bf5aa25e533ea7a082b592f0a1a12ffb. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by samsja** [comment]
Hey @karinazad the suggestion seems pretty reasonable, any chance you could describe a bit more what your end goal is so that we can think about the best way to solve the issue ? 

btw seems that the pr is also updating the uv.lock to use your internal packages repository, any chance you can uncommit the uv.lock change (its the reason most of the ci is failing)

**Comment by karinazad** [comment]
> Hey @karinazad the suggestion seems pretty reasonable, any chance you could describe a bit more what your end goal is so that we can think about the best way to solve the issue ?
> 
> btw seems that the pr is also updating the uv.lock to use your internal packages repository, any chance you can uncommit the uv.lock change (its the reason most of the ci is failing)

thanks @samsja!

I removed the uv.lock updates committed on accident

my main goal for step_hook is to support validation loss tracking + custom experiment metrics during training, without hardcoding project-specific logic into `train()`

custom metrics include e.g per-split validation perplexity

File: uv.lock [hunk 1]
```diff
@@ -3769,23 +3493,18 @@ wheels = [
 
 [[package]]
 name = "transformers"
-version = "4.57.6"
-source = { registry = "https://pypi.org/simple" }
+version = "5.2.0.dev0"
+source = { git = "https://github.com/huggingface/transformers.git?rev=609e3d5#609e3d585bfe2f78e95f18761dd85ae753da5f1b" }
```
[the remark this answers is not in the document]
Comment (karinazad) [reply uv.lock:3569]: removed uv.lock

File: uv.lock [hunk 2]
```diff
@@ -3995,8 +3727,8 @@ wheels = [
 
 [[package]]
 name = "vllm"
-version = "0.14.0"
-source = { registry = "https://pypi.org/simple" }
+version = "0.16.0"
+source = { url = "https://github.com/vllm-project/vllm/releases/download/v0.16.0/vllm-0.16.0-cp38-abi3-manylinux_2_31_x86_64.whl" }
```
[the remark this answers is not in the document]
Comment (karinazad) [reply uv.lock]: removed uv.lock

File: src/prime_rl/orchestrator/orchestrator.py [hunk 3]
```diff
@@ -729,7 +742,7 @@ def process_rollout(rollout: vf.RolloutOutput, rollout_idx: int) -> list[Trainin
                 evaluate_env(
                     env=eval_env,
                     env_name=eval_env_name,
-                    clients=inference_pool.clients,
+                    get_client=inference_pool.get_next_client,
```
[the remark this answers is not in the document]
Comment (karinazad) [reply src/prime_rl/orchestrator/orchestrator.py:779]: these changes were introduced during rebasing, updated now

File: src/prime_rl/orchestrator/config.py [hunk 4]
```diff
@@ -313,6 +313,17 @@ class EvalEnvConfig(EnvConfig):
...
+    # TODO: should live on the EnvConfig and also apply to training envs but
+    # this is hard right now because we use the vf.EnvGroup which treats all
+    # envs as one. for now training envs hardcode no retries, but we should
+    # probably treat them like environment groups long-term
+    max_retries: Annotated[
+        int,
+        Field(
+            description="Maximum number of times the environment will try to retry running a rollout.",
+        ),
+    ] = 0
```
[the remark this answers is not in the document]
Comment (karinazad) [reply src/prime_rl/orchestrator/config.py:324]: these changes were introduced during rebasing, updated now

**Comment by shriramc1** [comment]
Bumping this @samsja @karinazad!
